Enable virtualization support on Renesas RCAR platform

lorc opened the pull request:
There are two patches: 

One flushes TLB, to fix issue with page fault while writing to stack. This wasn't needed on QEMU, but is required on real HW.

Another moves platform data to `.nex` sections. Similar changes are required to other platforms, but I have no means to test them.

jforissier commented on core/arch/arm/plat-rcar/main.c:
```diff
@@ -53,7 +53,7 @@ register_dynamic_shm(NSEC_DDR_3_BASE, NSEC_DDR_3_SIZE);
 static void main_fiq(void);
 
-static const struct thread_handlers handlers = {
+static const struct thread_handlers handlers __nex_data = {
```
Isn't this needed for `plat-vexpress`, too? (I think you have tried virtualization with QEMUv8, right?).

lorc replied:
It is more complex than this. Yes, I initially did this on QEMUv8, but I had to enable gicv3.
Problem is that QEMU does not support virtualization extensions for gicv2, only for gicv3. So I had to rebuild both ARM-TF and OP-TEE with gicv3 enabled. 

This is why I didn't upstreamed changes to `plat-vexpress`. It requires multiple simultaneous changes in OP-TEE, ARM-TF and in build scripts. Also, I had to backport some virtualization-related patches to that version of QEMU that OP-TEE build scripts use.

jforissier commented:
Then that's fine. We can just say virt is not supported with qemu at the moment.
I asked because there is a `__nex_bss` in `plat-vexpress/main.c`.

lorc commented:
Let's say that virt support for qemu is partially done :) 
Those two `__nex_bss` entries in `plat-vexpress` are absolutely required. It is good to put `handlers` into `__nex_data` section as well, but actually it can work without this change. `handlers` are actually always read and never are written. "Golden" `.bss` section is mapped as RO when there is no virtual guest mapping active. So `handlers` are always accessible for reading. But it is inefficient to have them in `.bss` section, because every virtual guest context will contain copy of it. It is better to have only one copy in `.nex_data`. Hence this change.

Actually, if you are okay with moving QEMUv8 support to GICv3, I can prepare all needed changes. But probably I will need some help with ARM TF, because we need to maintain backward compatibility in some way.

Anyways, I added tags and rebased to master branch.

jforissier commented:
Thanks for the explanation.

About QEMUv8, it would certainly be good to have `CFG_VIRTUALIZATION` support because QEMU is super useful for development, debugging and testing. One step further, being able to run `xtest` with virtualization in the Travis CI environment would be cool, too! (`make check CFG_VIRTUALIZATION=y` more or less).

lorc commented:
Okay, I need to finish next version of path series for XEN and then I'll take a look how to properly add support for QEMU.

If you remember, I once had done that (https://github.com/OP-TEE/optee_os/pull/2370#issuecomment-397263614)  but it was not mature enough.

michpappas commented:
@lorc you probably remember that I brought up the issue with the TLB flush, during the review phase of the virtualization patches. Back then you arrogantly dismissed my comments as second-class and suggested that I opened PRs later. Now it turns out you individually fixed pretty much all of them. I don't want to make a big deal out of it, but I hope next time you'll have the courtesy to take them a bit more seriously.

CC: @jbech, @jenswi-linaro,  @jforissier

lorc commented:
@michpappas, yes, you surely mentioned TLB flush, but you comment was for `core_mmu_set_user_map()` function. At least this is what GitHub shows to me both in web UI and in emails. And I never saw any issues in this particular function. Actually, this function was in OP-TEE long before my PR and it is working very well.

As you can see my change was for `core_mmu_set_prtn()` function. Indeed, this newly added function had the issue with cache flushing. Sadly, you didn't pointed there. Again, maybe this is the issue of GitHub UI.

Anyways, I strongly disagree with your opinion about "arrogantly dismissed"  and "second-class comments". I never meant to discourage you in any way. I'm sorry that you see my comments in this way.

And thank you for your review for that PR.

jforissier commented:
Guys, those kinds of misunderstandings happen, please let's take it easy ;-)

@lorc and @michpappas, let's take a step back and look: you're both doing a great job with your contributions and reviewing work. As a maintainer I am pretty happy to see big patches coming from non-linaro members and I hope things will grow even more!

Thanks and please keep up the good work ;-)
